resources/dw/virtualwarehouse/impala/resource_impala_vw.go

```diff
@@ -0,0 +1,262 @@
+	vwID := state.ID.ValueStringPointer()
+
+	if clusterID == nil || vwID == nil {
+		resp.Diagnostics.AddError(
+			"Invalid State",
+			"ClusterID or Virtual Warehouse ID is missing.",
+		)
+		return
+	}
+
+	err := r.deleteVirtualWarehouse(ctx, clusterID, vwID)
+	if err != nil {
+		if strings.Contains(err.Error(), "Virtual Warehouse not found") {
+			return
+		}
+		resp.Diagnostics.AddError(
+			"Error deleting Impala Virtual Warehouse",
+			"Could not delete Impala Virtual Warehouse, unexpected error: "+err.Error(),
+		)
+		return
+	}
+
+	// Handle polling for not async call
+	pollingOptionsAvailable := state.PollingOptions != nil
+	isSynchronous := pollingOptionsAvailable && !state.PollingOptions.Async.ValueBool()
+	if isSynchronous {
+		err = r.pollForDeletion(ctx, state, clusterID, vwID)
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Error waiting for Impala Virtual Warehouse deletion",
+				"Could not delete Impala Virtual Warehouse, unexpected error: "+err.Error(),
+			)
+		}
+	}
+}
+
+func (r *impalaResource) stateRefresh(ctx context.Context, clusterID *string, vwID *string, callFailedCount *int, callFailureThreshold int) func() (any, string, error) {
+	return func() (any, string, error) {
+		tflog.Debug(ctx, "About to describe Impala")
+		params := operations.NewDescribeVwParamsWithContext(ctx).
+			WithInput(&models.DescribeVwRequest{ClusterID: clusterID, VwID: vwID})
+		resp, err := r.client.Dw.Operations.DescribeVw(params)
+		if err != nil {
+			if strings.Contains(err.Error(), "Virtual Warehouse not found") {
+				return &models.DescribeVwResponse{}, "Deleted", nil
+			}
+			*callFailedCount++
+			if *callFailedCount <= callFailureThreshold {
+				tflog.Warn(ctx, fmt.Sprintf("could not describe Data Warehouse Impala Virtual Warehouse "+
+					"due to [%s] but threshold limit is not reached yet (%d out of %d).", err.Error(), callFailedCount, callFailureThreshold))
+				return nil, "", nil
+			}
+			tflog.Error(ctx, fmt.Sprintf("error describing Data Warehouse Impala Virtual Warehouse due to [%s] "+
+				"failure threshold limit exceeded.", err.Error()))
+			return nil, "", err
+		}
+		*callFailedCount = 0
+		vw := resp.GetPayload()
+		tflog.Debug(ctx, fmt.Sprintf("Described Impala %s with status %s", vw.Vw.ID, vw.Vw.Status))
+		return vw, vw.Vw.Status, nil
+	}
+}
+
+func createVwRequestFromPlan(plan *resourceModel) *models.CreateVwRequest {
+	req := &models.CreateVwRequest{
+		Name:      plan.Name.ValueStringPointer(),
+		ClusterID: plan.ClusterID.ValueStringPointer(),
+		DbcID:     plan.DatabaseCatalogID.ValueStringPointer(),
+		VwType:    models.VwTypeImpala.Pointer(),
+	}
+	if imageVersion := plan.ImageVersion.ValueString(); imageVersion != "" {
+		req.ImageVersion = imageVersion
+	}
+	return req
+}
+
+func (r *impalaResource) waitForVwRunning(ctx context.Context, plan *resourceModel, payload *models.CreateVwResponse) error {
+	clusterID := plan.ClusterID.ValueStringPointer()
+	vwID := &payload.VwID
```

Comment:
You only need the payload to get the VwID. You should always try to pass the minimum required objects to a function, if you need the VwID pass the VwID, if you need many properties, it is ok to pass the payload.

Reply:
Done